### PR TITLE
scheduler: Implement custom RunScheduler

### LINKED ISSUE #6: Celerybeat custom scheduler
- [Celery Periodic Tasks custom scheduler](https://docs.celeryq.dev/en/latest/userguide/periodic-tasks.html#using-custom-scheduler-classes), that could hook into the `Job` model (or possibly a completely new model that `Job` references).
- We can look into the existing `django-celery-beat` library, especially in the logic from [`schedulers.py`](https://github.com/celery/django-celery-beat/blob/main/django_celery_beat/schedulers.py) and [`models.py`](https://github.com/celery/django-celery-beat/blob/main/django_celery_beat/models.py)

### PR DESCRIPTION
**Pull request by yashlamba**
Closes #6

Depends on: https://github.com/inveniosoftware/invenio-records-resources/pull/579

### CODE REVIEW

**Review by slint — APPROVED** [review]
LGTM, some minor comments

File: invenio_jobs/models.py [hunk 1]
```diff
@@ -40,12 +41,20 @@ class Job(db.Model, Timestamp):
     default_args = db.Column(JSON, default=lambda: dict(), nullable=True)
     schedule = db.Column(JSON, nullable=True)
 
+    last_run_at = db.Column(db.DateTime, nullable=True)
```
[the remark this answers is not in the document]
Comment (yashlamba) [reply invenio_jobs/models.py:44]: In celery terms, our schedules are coming from Jobs, which maintain the last_run_at. We can compute it, but it still has to come from the Job model

and at this stage, the run hasn't been created yet, ideally, the latest run's create time == Job.last_run_at.

[the remark this answers is not in the document]
Comment (slint) [reply invenio_jobs/models.py:44]: What I don't understand is why does Celery care where the information comes from, i.e. is it part of an interface/class that we pass? Let's discuss IRL maybe

File: invenio_jobs/services/scheduler.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    @classmethod
+    def from_job(cls, job):
+        return cls(
+            job=job,
+            name=job.title,
+            schedule=job.parsed_schedule,
+            args=job.default_args,
+            task=job.task,
+            kwargs={},
```
Comment (slint) [inline invenio_jobs/services/scheduler.py:34]: minor: we'll actually have to split `default_args` into `args` and `kwargs` (it matters for Celery tasks). The best way to do so at the moment is by inspecting the task functions (i.e. what's done in the `models.Task.signature` property). I would find a couple of example tasks and see if these splitting logic can be consistently replicated.

File: invenio_jobs/utils.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        setattr(run, kw, value)
+    db.session.commit()
+
+
+def is_a_job(func):
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        # TODO Add failsafe of if run doesn't exist
+        run = Run.query.filter_by(task_id=self.request.id).one_or_none()
+        update_run(run, status=RunStatusEnum.RUNNING, started_at=datetime.now())
```
Comment (slint) [inline invenio_jobs/utils.py:24]:
```suggestion
        update_run(run, status=RunStatusEnum.RUNNING, started_at=datetime.utcnow())
```

Ditto for other places where we take timestamps.

File: invenio_jobs/services/scheduler.py (hunk 2 above)
Comment (yashlamba) [reply invenio_jobs/services/scheduler.py:34]: So we are going with all arguments as kwargs as we don't use positional only arguments as of now. Might be implemented later.

**Comment by yashlamba** [comment]
Rebased on top of #14

File: invenio_jobs/services/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                )
+            else:
+                if result and hasattr(result, "id"):
+                    logger.debug("%s sent. id->%s", entry.task, result.id)
+                else:
+                    logger.debug("%s sent.", entry.task)
+
+    def sync(self):
+        """Sync Jobs from db to the scheduler."""
+        # TODO Should we also have a cleaup task for runs? "stale" run (status running, starttime > hour, Run pending for > 1 hr)
```
Comment (alejandromumo) [inline invenio_jobs/services/scheduler.py:93]: should we move this `TODO` to a new ticket?

File: invenio_jobs/services/scheduler.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        with self.app.flask_app.app_context():
+            jobs = Job.query.filter(Job.active == True).all()
+            self.entries = {}  # because some jobs might be deactivated
```
Comment (alejandromumo) [inline invenio_jobs/services/scheduler.py:96]: what does it mean `some jobs might be deactivated`? Doesn't the `RunScheduler`object have the entries initialized after instantiation?

File: invenio_jobs/tasks.py [hunk 6]
```diff
@@ -6,11 +6,51 @@
...
+from invenio_db import db
 
+from invenio_jobs.models import Run, RunStatusEnum
 
-@shared_task(ignore_result=True)
-def run_job(job_id, args=None):
-    """Run a job."""
-    pass
+
+# TODO 1. Move to service? 2. Don't use kwargs?
```
Comment (alejandromumo) [inline invenio_jobs/tasks.py:17]: remove `TODO` if fixed, otherwise we can create tickets.

File: invenio_jobs/tasks.py [hunk 7]
```diff
@@ -6,11 +6,51 @@
...
+def execute_run(self, run_id, kwargs=None):
+    """Execute and manage a run state and task."""
+    run = Run.query.filter_by(id=run_id).one_or_none()
+    task = self.app.tasks.get(run.job.task)
+    update_run(run, status=RunStatusEnum.RUNNING, started_at=datetime.now(timezone.utc))
+
+    try:
+        result = task.apply(kwargs=kwargs)
+    except Exception as e:
+        # TODO should we log the error in message?
```
Comment (alejandromumo) [inline invenio_jobs/tasks.py:37]: If you need to log this on sentry for visibility :

```python
current_app.logger.exception("...")
```

Comment (alejandromumo) [reply invenio_jobs/tasks.py:37]: we agreed on creating a ticket and removing  the TODO

Comment (yashlamba) [reply invenio_jobs/tasks.py:37]: https://github.com/inveniosoftware/invenio-jobs/issues/22

File: invenio_jobs/services/scheduler.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        Scheduler.__init__(self, *args, **kwargs)
+
+    @property
+    def schedule(self):
+        """Get currently scheduled entries."""
+        return self.entries
+
+    def setup_schedule(self):
+        """Setup schedule."""
+        # TODO Check whether we need the celery backend task?
```
[the remark this answers is not in the document]
Comment (alejandromumo) [reply invenio_jobs/services/scheduler.py:60]: agreed on creating a ticket and remove the `TODO`

File: invenio_jobs/services/scheduler.py (hunk 4 above)
Comment (alejandromumo) [reply invenio_jobs/services/scheduler.py:93]: we agreed on leaving the `TODO` here but also open a ticket

File: invenio_jobs/tasks.py (hunk 6 above)
Comment (alejandromumo) [reply invenio_jobs/tasks.py:17]: agreed on creating a ticket and checking later on

**Pull request merged**